### PR TITLE
  Add VLM support to DynamicInferenceEngine

### PR DESCRIPTION
**Pull request by RPrenger**
Extend the dynamic-batching inference path to accept per-request image
  inputs. The engine now offers VLM kwargs on add_request (imgs,
  imgs_sizes, num_tiles, num_img_embeddings_per_tile) that:

    * expand <image> placeholders into pad tokens (static-resolution and dynamic-resolution cases both supported);
    * run the vision encoder on the first PP stage to produce per-request image embeddings;
    * attach the embeddings and image-token mask to the context so the decoder forward can splice them back in.

  DynamicInferenceContext gains _request_to_image_embeddings/
  _request_to_image_token_mask/_request_to_image_token_count dicts and
  four new accessors (add_vlm_request_data, remove_vlm_request_data,
  current_image_token_mask, current_image_embeddings); the per-request
  data is cleared at reset_metadata and on request completion in
  post_process_requests.

  VLMInferenceWrapper gains expand_image_tokens, _forward_vision_encoder
  and _forward_dynamic to back the engine path. All preexisting wrapper
  methods are unchanged, and the new behavior is dormant unless a caller
  passes imgs/imgs_sizes/num_tiles to add_request, so text-only inference
  is unaffected.

  TextGenerationController._dynamic_step_forward_logits picks up the
  per-request image data from the context and forwards it through the
  wrapper's inference_input dict. Out-of-vocabulary placeholders
  introduced by VLM expansion would otherwise leak into expert-parallel
  dummy forwards (whose logits are discarded but whose token IDs still
  have to be in vocab); _sanitize_dummy_input_ids replaces them with a
  stable valid token on the dummy path only.

  DynamicVLMInferenceRequest combines DynamicInferenceRequest and
  VLMInferenceRequest, carrying the precomputed image_embeddings and
  image_token_mask alongside the usual request state.

- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/contexts/dynamic_context.py [hunk 1]
```diff
@@ -4137,3 +3990,113 @@ def get_kvcache_utilization_stats(self) -> dict:
...
+                -1,
+                dtype=torch.long,
+                device=torch.cuda.current_device(),
+            )
+
+        segments: List[Tensor] = []
+        cumulative_offset = 0
+        for row_idx in range(self.paused_request_count, self.total_request_count):
+            request_id = int(self.request_ids[row_idx].item())
+            query_len = int(self.request_query_lengths[row_idx].item())
```
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py]: Instead of doing individual `.item` calls for each `row_idx`, can we do batched transfers of `request_ids[self.paused_request_count : self.total_request_count]` and `self.request_query_lengths[self.paused_request_count : self.total_request_count]` before the loop?

File: megatron/core/inference/contexts/dynamic_context.py [hunk 2]
```diff
@@ -4137,3 +3990,113 @@ def get_kvcache_utilization_stats(self) -> dict:
...
+        unsqueezed to [sum(seq_img_i), 1, hidden]. The downstream consumer
+        does ``permute(1, 0, 2).reshape(-1, hidden)`` to yield the flat
+        [total_image_tokens, hidden] array indexed by
+        :meth:`current_image_token_mask`.
+
+        Returns tensor of shape [total_image_tokens, 1, hidden] or None.
+        """
+        parts: List[Tensor] = []
+        for row_idx in range(self.paused_request_count, self.total_request_count):
+            request_id = int(self.request_ids[row_idx].item())
```
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py]: Same comment about batched transfer on `request_ids`

File: megatron/core/inference/engines/dynamic_engine.py [hunk 3]
```diff
@@ -1174,16 +1154,84 @@ def add_request(
...
 
-        # Initialize request.
-        request = DynamicInferenceRequest(
-            request_id=request_id,
-            prompt=prompt_str,
-            prompt_tokens=tokens,
-            sampling_params=sampling_params,
-            block_size_tokens=self.context.block_size_tokens,
-            enable_prefix_caching=self.context.enable_prefix_caching,
+        # --- VLM: expand image tokens and run vision encoder ---
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py]: Can we move the image-related code to an internal helper function? 
```
if has_images:
    request = self._build_vlm_request(imgs, num_tiles, imgs_sizes)
else:
    request = DynamicInferenceRequest(...)
```

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 4]
```diff
@@ -705,25 +616,66 @@ def _dynamic_step_context_init(
...
+        Expert-parallel dummy forwards are synthetic; they only keep ranks with
+        no local request participating in the same collectives as active ranks.
+        Their logits are discarded, so any out-of-vocabulary placeholder token
+        can be replaced with a stable valid token without affecting user-visible
+        output. This keeps the dummy path independent of how dynamic VLM
+        resolution expands image placeholders for real requests.
+        """
+        invalid_mask = (input_ids < 0) | (input_ids >= self.vocab_size)
+        if not invalid_mask.any():
+            return input_ids
```
Comment (santhnm2) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py:721]: Can we just directly do `input_ids[invalid_mask] = 0`? This should be a no-op if `invalid_mask.any()` is False right?

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 5]
```diff
@@ -705,25 +616,66 @@ def _dynamic_step_context_init(
...
+
+        input_ids = input_ids.clone()
```
Comment (santhnm2) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py:723]: Why do we need the `.clone()`?

File: megatron/core/inference/contexts/dynamic_context.py (hunk 1 above)
Comment (RPrenger) [reply megatron/core/inference/contexts/dynamic_context.py]: Yes, I pushed a change.  Though I don't think these request_ids are on the CPU, so `.item()` isn't causing a GPU sync or anything right now.  But probably good to avoid it.

File: megatron/core/inference/contexts/dynamic_context.py (hunk 2 above)
Comment (RPrenger) [reply megatron/core/inference/contexts/dynamic_context.py]: Same as the last comment.  Fixed now.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 3 above)
Comment (RPrenger) [reply megatron/core/inference/engines/dynamic_engine.py]: Yep.  Changed it.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py (hunk 4 above)
Comment (RPrenger) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py:721]: So `input_ids` is a view into `context.gpu_view.token_to_input_ids` and I think it's cached for reuse across steps in `context._input_position_views`.  So we need the `.clone()` I think.  And then the `.any()` is so we can skip both `.clone()` and the assign if there are no invalid tokens.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py (hunk 5 above)
Comment (RPrenger) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py:723]: See above.  I think we need the clone because it gets cached and reused.

Comment (santhnm2) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py:723]: I see, could you add a comment explaining this?

**Review by santhnm2 — APPROVED** [review]

**Comment by RPrenger** [comment]
To keep this PR reviewable, it intentionally ships engine + wire plumbing for VLM dynamic batching without the LLaVAModel companion changes

(forward_lm_only, _dynamic_resolution / _patch_dim / _class_token_len), which will land
in a follow-up PR. Text-only inference is unaffected.

Latest commit addresses each comment:

- #12 — _forward_dynamic now raises NotImplementedError with an explicit "requires
companion LLaVAModel changes" message when forward_lm_only isn't present, instead of an
AttributeError at call time.
- #13 — expand_image_tokens and _forward_vision_encoder now raise NotImplementedError
when a caller passes imgs_sizes (dynamic resolution) but the model lacks
_dynamic_resolution, instead of silently falling through to the static path. Also removed
the unused drop_class_token / class_token_len locals inside the dynamic branch.
- #14 — _build_vlm_request asserts pipeline-parallel world size 1 (PP>1 needs the
non-first-stage embedding recv path that the wrapper's _recv_only_vision_embeds TODO
refers to; not upstream yet). Added the shape assertion you asked for at the splice site
so token-count / embedding-count disagreement fails loudly rather than indexing out of
bounds.
- #15 — Both current_image_token_mask and current_image_embeddings now short-circuit and
return None when the per-request VLM dicts are empty. Text-only workloads skip the
.tolist() / torch.cat on every step.

**Comment by kvareddy** [comment]
@lmcafee-nvidia can you please take a look at this MR?

**Comment by kvareddy** [comment]
@RPrenger  can we add unit test cases to Megatron Inference for VLM support?

**Comment by RPrenger** [comment]
> @RPrenger can we add unit test cases to Megatron Inference for VLM support?

Happy to add tests, but I'd like to propose deferring them to the follow-up PR that lands the `LLaVAModel` companion changes (`forward_lm_only`, `_dynamic_resolution` etc.). Rationale:

This PR ships engine + wire plumbing only. Every new code path either (a) fires only when the `LLaVAModel` companion is present or (b) is a `NotImplementedError` guard for the case where it isn't. Tests I could write against upstream today would look like:

- Pure `DynamicInferenceContext` mask/embedding accessor tests with hand-built tensors (verifies the perf early-return + offset math).
- `_sanitize_dummy_input_ids` tensor tests (out-of-vocab → 0 replacement).
- `NotImplementedError` guards fire when `forward_lm_only` / `_dynamic_resolution` are absent.
- `expand_image_tokens` static path with a `MagicMock` module.

Those exercise the plumbing but don't validate any actual VLM inference — the mocks stand in for the very code that isn't upstream yet. A follow-up PR with the `LLaVAModel` companion + a real end-to-end VLM inference test would be a stronger safeguard against regressions than mock-heavy plumbing tests here.

If you'd prefer we get some coverage in this PR anyway, happy to add the guard-and-context tests above — just wanted to check on scope first.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 6]
```diff
@@ -1189,19 +1215,111 @@ def add_request(
...
+            request_id, image_embeddings=image_embeddings, image_token_mask=mask_tensor
+        )
+
+        return DynamicVLMInferenceRequest(
             request_id=request_id,
             prompt=prompt_str,
             prompt_tokens=tokens,
             sampling_params=sampling_params,
             block_size_tokens=self.context.block_size_tokens,
             enable_prefix_caching=self.context.enable_prefix_caching,
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py:1314]: I could be wrong, but it seems placeholders (-1 values) are initially used for the image tokens until the first forward pass runs. this means that the placeholders are what's added to the request object, and added to the prefix cache if we are using prefix caching. since the cache sees the -1's and not the actual image, then multiple requests can point to the same cache blocks. for example, “describe <image>” with first a cat, and later a dog, the dog could reuse the cat’s saved state for the dog request. if this is the case, then we should disable prefix caching for image data; or if possible maybe we could cache only the text tokens before the first image.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 7]
```diff
@@ -4137,3 +4147,127 @@ def get_kvcache_utilization_stats(self) -> dict:
...
+        segments: List[Tensor] = []
+        cumulative_offset = 0
+        for request_id, query_len in zip(active_request_ids, active_query_lengths):
+            per_request_mask = self._request_to_image_token_mask.get(request_id, None)
+            if per_request_mask is None:
+                seg = torch.full(
+                    (query_len,), -1, dtype=torch.long, device=torch.cuda.current_device()
+                )
+            else:
+                seg = per_request_mask[:query_len].clone()
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py:4222]: this always reads image positions from the start of the prompt, even when processing a later chunk. for example, if the first chunk is `[image, image, "what"]` and the second is `["is", "this", "?"]`, the second chunk can reuse the first chunk’s image positions and replace `"is"` and `"this"` with image data. I think the image positions need to be aligned with the chunk currently being processed, and probably add a test for this case.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 8]
```diff
@@ -4137,3 +4147,127 @@ def get_kvcache_utilization_stats(self) -> dict:
...
+
+        active_request_ids = self.request_ids[
+            self.paused_request_count : self.total_request_count
+        ].tolist()
+
+        parts: List[Tensor] = []
+        for request_id in active_request_ids:
+            emb = self._request_to_image_embeddings.get(request_id, None)
+            if emb is not None:
+                parts.append(emb.reshape(-1, emb.shape[-1]))
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py:4270]: I believe this combines a fixed-size, multi-tile image in the wrong order. for example, if tile A has patches `A1, A2` and tile B has `B1, B2`, the expected order is `A1, A2, B1, B2`, but this produces `A1, B1, A2, B2`. the patches are then attached to the wrong image positions. we should probably put all patches from one tile together before flattening the data, and add a test for this.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 9]
```diff
@@ -2534,6 +2539,11 @@ def reset_metadata(self) -> None:
             token_count=0, prefill_req_count=0, decode_req_count=0
         )
 
+        # Reset VLM data.
+        self._request_to_image_embeddings.clear()
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py:2543]: in `RECOMPUTE` pause/resume mode, this reset erases the saved image positions and image data, but resume restores only the ordinary request information. for example, if a half-finished photo request is paused to free memory and then resumed, its reserved image positions remain but the actual photo data is gone. the request can then fail or continue without using the photo. we should preserve or restore the image data for every resumed request, or reject pause/resume while image requests are present. also a good idea to add a test for this.

File: megatron/core/inference/engines/dynamic_engine.py
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py:1]: please add tests for the entire PR.

Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py:1]: also, we should add support for the client/server. I don't think the client currently has any way of sending image data.
